File: src/staleness_consumer_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <thread>
#include <chrono>

ros::Subscriber img_sub;
int num_samples, required_samples;
std::vector<double> samples;

static void calcStatistics();

static void image_callback(const sensor_msgs::ImageConstPtr& img_msg)
{
    auto now = ros::Time::now();
    auto staleness = (now - img_msg->header.stamp).toSec();
    samples.push_back(staleness);
    if (samples.size() >= required_samples)
    {
        calcStatistics();
        std::exit(0);
    }
    ROS_INFO_STREAM("Processing message with staleness " << staleness);
    std::this_thread::sleep_for(std::chrono::seconds(1));
}

static void calcStatistics()
{
    double average = 0.0;
    for (const auto& sample : samples)
    {
        average += sample;
    }
    average /= samples.size();
    double sq_sum = 0.0;
    for (const auto& sample: samples)
    {
        sq_sum += (sample - average) * (sample - average);
    }
    double stdev = std::sqrt(sq_sum / samples.size());

    ROS_INFO_STREAM("Average staleness: " << average << "; standard deviation: " << stdev << " (based on " << samples.size() << " samples)");
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "staleness_consumer");
    ros::NodeHandle nh, nh_priv("~");
    img_sub = nh.subscribe<sensor_msgs::Image>("/staleness_producer/image", 1, &image_callback);
    required_samples = nh_priv.param<int>("required_samples", 10);
    samples.reserve(required_samples);

    ros::spin();
    return 0;
}

File: scripts/staleness_consumer_node.py
#!/usr/bin/env python
import rospy
from sensor_msgs.msg import Image

rospy.init_node('staleness_consumer')

samples = []
required_samples = rospy.get_param("~required_samples", default=10)
buffer_size = rospy.get_param("~buffer_size", rospy.topics.DEFAULT_BUFF_SIZE)


def image_callback(msg):
    staleness = (rospy.Time.now() - msg.header.stamp).to_sec()
    samples.append(staleness)
    if len(samples) >= required_samples:
        calcStatistics()
        rospy.signal_shutdown('enough samples gathered')
    rospy.loginfo("Processing message with staleness {}".format(staleness))
    rospy.sleep(1.0)


def calcStatistics():
    average = sum(samples) / len(samples)
    sq_sum = sum((s - average) ** 2 for s in samples)
    stdev = (sq_sum / len(samples)) ** 0.5
    rospy.loginfo("Average staleness: {}; standard deviation: {} (based on {} samples)".format(
        average, stdev, len(samples)
    ))


rospy.loginfo("Running with buff_size={}".format(buffer_size))
img_sub = rospy.Subscriber('/staleness_producer/image', Image, image_callback, queue_size=1, buff_size=buffer_size)
rospy.spin()
